internal/prototest/yaml.go

```diff
@@ -0,0 +1,370 @@
+		return d
+
+	case protoreflect.EnumNumber:
+		enum := f.Enum()
+		if value := enum.Values().ByNumber(v); value != nil {
+			return value.Name()
+		} else {
+			return int32(v)
+		}
+
+	case []byte:
+		return string(v)
+
+	default:
+		return v
+	}
+}
+
+// write writes a value returned by [toYAML.value] into the internal output
+// buffer.
+func (y *toYAML) write(v any) {
+	switch v := v.(type) {
+	case int32, int64, uint32, uint64, float32, float64, protoreflect.Name:
+		fmt.Fprint(&y.out, v)
+	case string:
+		fmt.Fprintf(&y.out, "%q", v)
+	case *doc:
+		isOneLine := func(v any) bool {
+			maxWidth := y.MaxWidth
+			if maxWidth == 0 {
+				maxWidth = 80
+			}
+			maxWidth -= y.nesting * 2
+
+			doc, ok := v.(*doc)
+			return !ok || doc.width < maxWidth
+		}
+
+		if isOneLine(v) {
+			switch {
+			case v.isArray:
+				y.out.WriteString("[")
+				for i, pair := range v.pairs {
+					if i > 0 {
+						y.out.WriteString(", ")
+					}
+					y.write(pair[1])
+				}
+				y.out.WriteString("]")
+
+			case len(v.pairs) == 0:
+				y.out.WriteString("{}")
+
+			default:
+				// Special case: if we are a list element, and there is only
+				// one entry, print it directly.
+				if len(v.pairs) == 1 && strings.HasSuffix(y.out.String(), "- ") {
+					y.write(v.pairs[0][0])
+					y.out.WriteString(": ")
+					y.write(v.pairs[0][1])
+					return
+				}
+
+				y.out.WriteString("{ ")
+				for i, pair := range v.pairs {
+					if i > 0 {
+						y.out.WriteString(", ")
+					}
+					y.write(pair[0])
+					y.out.WriteString(": ")
+					y.write(pair[1])
+				}
+				y.out.WriteString(" }")
+			}
+			return
+		}
+
+		for _, pair := range v.pairs {
+			oneLine := isOneLine(pair[1])
+			y.indent()
+
+			if pair[0] == nil {
+				y.out.WriteString("- ")
+			} else {
+				y.write(pair[0])
+				if !oneLine {
+					y.out.WriteString(":\n")
+				} else {
+					y.out.WriteString(": ")
+				}
+			}
+
+			if !oneLine {
+				y.nesting++
+			}
+
+			y.write(pair[1])
+			if !oneLine {
+				y.nesting--
+			} else {
+				y.out.WriteString("\n")
+			}
+		}
+	}
+}
+
+// indent appends indentation if necessary.
+func (y *toYAML) indent() {
+	s := y.out.String()
+	if s == "" || strings.HasSuffix(s, "\n") {
+		for i := 0; i < y.nesting; i++ {
+			y.out.WriteString("  ")
+		}
+	}
+}
+
+// doc is a generic document structure used as an intermediate for generating
+// the compressed output of ToYAML.
+//
+// It is composed of an array of pairs of arbitrary values.
+type doc struct {
+	pairs [][2]any
+
+	width              int
+	isArray, needsSort bool
+}
+
+// push adds a new entry to this document.
+//
+// All pushes entries must either have a non-nil key OR a nil key.
+func (d *doc) push(k, v any) {
+	if len(d.pairs) == 0 {
+		d.isArray = k == nil
+	} else if d.isArray != (k == nil) {
+		panic("misuse of doc.push()")
+	}
+
+	d.pairs = append(d.pairs, [2]any{k, v})
+}
+
+// compress prepares a document for printing by compressing elements as
+// appropriate.
+func (d *doc) compress() {
```

Comment:
nit: since it also measures the size/width of the document, maybe `prepare` is a better name.